### PR TITLE
Fix: only parameterize when dynamic values are used

### LINKED ISSUE #21: Fix parametric interpolation in Quil transpilation
Re: https://github.com/rigetti/qcs-sdk-qir/pull/16#discussion_r838799023

### PR DESCRIPTION
**Pull request by nilslice**
Fixes #21.

### CODE REVIEW

File: src/transform/shot_count_block/pattern.rs [hunk 1]
```diff
@@ -375,6 +383,11 @@ fn get_quil_parameter_expression<'ctx>(
     float_value: FloatValue<'ctx>,
 ) -> Expression {
     let index = get_quil_parameter_index(pattern_context, float_value);
+    if float_value.is_const() {
+        let constant = float_value.get_constant().unwrap().0;
```
Comment (nilslice) [inline src/transform/shot_count_block/pattern.rs:387]: Any thoughts on how this `unwrap` might fail and what to actually do in the event it does are very welcome. I think I'll change it at least to be `expect`, but would be glad to do some better edge-case handling however possible.

File: src/transform/shot_count_block/pattern.rs [hunk 2]
```diff
@@ -192,6 +192,14 @@ impl<'ctx> ShotCountPatternMatchContext<'ctx> {
 
         Ok(pattern_context)
     }
+
+    /// Returns the parameters which do not have a constant value.
+    pub(crate) fn dynamic_parameters(&self) -> Vec<&FloatValue<'ctx>> {
```
Comment (kalzoo) [inline src/transform/shot_count_block/pattern.rs:197]:
```suggestion
    pub(crate) fn get_dynamic_parameters(&self) -> Vec<&FloatValue<'ctx>> {
```

Just a style opinion given that this allocates a new vec and is not a property

File: src/transform/shot_count_block/pattern.rs [hunk 3]
```diff
@@ -375,6 +383,14 @@ fn get_quil_parameter_expression<'ctx>(
     float_value: FloatValue<'ctx>,
 ) -> Expression {
```
Comment (kalzoo) [inline src/transform/shot_count_block/pattern.rs:384]: How about returning an `Option<Expression>` to avoid the panic?

**Review by kalzoo — APPROVED** [review]
Well done, @nilslice ! A neat fix and good tests

File: src/transform/shot_count_block/pattern.rs [hunk 4]
```diff
@@ -375,6 +383,14 @@ fn get_quil_parameter_expression<'ctx>(
...
     let index = get_quil_parameter_index(pattern_context, float_value);
+    if float_value.is_const() {
+        let constant = float_value
+            .get_constant()
+            .expect("parameter constant should exist")
+            .0;
+        return Expression::Number(constant.into());
+    }
```
Comment (kalzoo) [inline src/transform/shot_count_block/pattern.rs:392]:
```suggestion
    if let Some(constant) = float_value.get_constant() {
        return Expression::Number(constant.0.into());
    }
```

File: src/transform/shot_count_block/pattern.rs (hunk 3 above)
Comment (kalzoo) [reply src/transform/shot_count_block/pattern.rs:384]: never mind - better suggestion below

Comment (nilslice) [reply src/transform/shot_count_block/pattern.rs:384]: excellent - thank you!

**Pull request merged**